### PR TITLE
feat: Automate helm docs

### PR DESCRIPTION
**Pull request by souleb**
This PR is about automating the helm docs generation for a given helm chart version. It relies on [helm-docs](https://pkg.go.dev/github.com/norwoodj/helm-docs/pkg/helm).

A template is provided in order to generate `heml.rst`, the reStructuredText formatted doc. This template has access to all the metadata in the chart's `requirements.yaml`, and `Chart.yaml` as well as the `chart values`. The advantage here is that any changes to those special files in the chart or in the template itself, can automatically generate an up to date documentation.

**Note** As we want to avoid any manual action i.e. copy for the helm chart `values`, the `values.yaml` is the source of truth  and has to be update to contain all needed information. One critical missing part is the fields descriptions. We can also set markers in it in order to hide some fields.

## TO-DO
- [x] Create a Makefile to install `helm-docs`, fetch the right chart version and generate `helm.rst`
- [x] Provide a `helm.rst.gotmpl` that can generate the right `helm.rst`
- [x] Update the github workflows in order to automatically regenerate the `helm.rst` before publishing a new version
- [x] Update the helm chart to hold the fields descriptions

Edit: Depends on https://github.com/Mellanox/network-operator/pull/1011

### CODE REVIEW

**Comment by adrianchiris** [comment]
some feedback:

we should generate helm rst doc but should not blindly publish it on merge in workflow, it can easily go south when autogenerating from comments added by devs, without any review in the middle this will most likely get missed (or caught too late)


1. having a generate-helm-doc makefile target to generate doc from helm chart  + template
2. this should be integrated into network operator CD pipeline (after the chart was generated and pushed, or provided via path to local chart). ending up with a PR to this repo which updates helm rst doc
3. helm update PR is reviewed and merged like any other doc change
4. eventually  a release is tagged in networo-operator-docs proj.

**Comment by adrianchiris** [comment]
> Thanks @adrianchiris for the feedback.
> 
> > we should generate helm rst doc but should not blindly publish it on merge in workflow, it can easily go south when autogenerating from comments added by devs, without any review in the middle this will most likely get missed (or caught too late)
> 
> I agree with your concern here. But I think we should always regenerate the `helm.rst` before publishing to make sure we have an up to date version. Otherwise, editing manually after merging a PR that update the `helm.rst` doc can still leave it out of sync. Also, I think that the `helm.rst` should live in the same repo that generates it to have a simpler workflow.
> 
> So what about this update to your list? :
> 
> 1. having a generate-helm-doc makefile target to generate doc from helm chart + template
> 2. **A makefile target that runs on every PR to regenerate the doc(`helm.rst`). This makes sure that the doc is part of the PR review, and it should only change if the chart itself is updated.**
> 3. **This should be integrated into network operator repository. So `helm.rst` will be part of the network operator release tarball. We avoid creating PRs to the network-operator-docs repo, and have only a single file to track for this.**
> 4. eventually a release is tagged in network-operator-docs repo. **As part of the release pipeline workflow, the `helm.rst` is pulled from the corresponding network operator release**

Hey @souleb ! Here are some point i had in mind (which i guess we disagree on :) ):

- i see network-operator project as a source of "metadata" used for generating docs in this repo (e.g component version, helm chart comments etc...). doc artefacts remain here.
- i see our release (CD) pipeline updating documentation for helm chart and versions when releasing new version of network operator (beta, rc , GA...) by: running some makefile target (e.g make generate-helm-doc) then submitting PR to this project to be reviewed and merged. (manual fixes may be needed both here and in the chart until we dial it in)
- once we see its all working well, CD pipeline can push directly (or just merge it own PR)

some of my comments reviewing this PR are related to the points above

File: Makefile
Comment (adrianchiris) [inline Makefile:19]: can we have helm-docs as a separate target only ?

**Comment by souleb** [comment]
> - i see our release (CD) pipeline updating documentation for helm chart and versions when releasing new version of network operator (beta, rc , GA...) by: running some makefile target (e.g make generate-helm-doc) then submitting PR to this project to be reviewed and merged. (manual fixes may be needed both here and in the chart until we dial it in)
> - once we see its all working well, CD pipeline can push directly (or just merge it own PR)

So every time a developer makes change to the api and the helm chart in `network-operator`, he will have to deal with 2 PRs, the fist in `network-operator` and a second in `network-operator-docs`? If I understand, the second PR will be generated once the first is merged. So if there is an issue, the developer will have to resubmit a PR in `network-operator` to fix it.

> - as a developer writing docs for a specific feature i dont want my PR to contain distracting information (e.g updates to helm.rst)

This is not distracting, because this will be part of the PR only if there are changes in the helm chart or doc template? And this way we have the fastest feedback loop and can avoid back and forth between 2 repositories.

File: templates/helm.rst.gotmpl
Comment (souleb) [inline templates/helm.rst.gotmpl:158]: I believe this is not true anymore... Those fields are commented in the `helm chart`.

**Comment by souleb** [comment]
Do not merge before https://github.com/Mellanox/network-operator/pull/1011

File: docs/customizations/helm.rst
Comment (rollandf) [inline docs/customizations/helm.rst:181]: Is it supposed to be skipped?

File: docs/customizations/helm.rst
Comment (rollandf) [inline docs/customizations/helm.rst:89]: These are tough to read, and not really useful

File: docs/customizations/helm.rst
Comment (souleb) [reply docs/customizations/helm.rst:181]: It will be skipped, once we merge [1011](https://github.com/Mellanox/network-operator/pull/1011)

File: docs/customizations/helm.rst
Comment (souleb) [reply docs/customizations/helm.rst:89]: This is not the final `helm.rst`. It will be rendered properly once the depended PR is merged. If you want a preview look at https://github.com/Mellanox/network-operator/blob/6ae08a018f019d12fc13d4630209e25cd22cc5ef/docs/charts/helm.rst.

**Review by rollandf — APPROVED** [review]

File: Makefile
Comment (rollandf) [inline Makefile:48]: Can we have an option to run it on a directory in addition to the option of downloading the chart?

Comment (souleb) [reply Makefile]: I have added separate target for branch or ngc.

File: Makefile
Comment (adrianchiris) [inline Makefile:22]: just making sure:

if i run
`HELM_CHART_DEP_ROOT=/path/to/my/local/network-operator/deployment/network-operator make helm-docs`

it will work right ?

Comment (souleb) [reply Makefile:22]: yes, but there are two assumptions here:
- `HELM_CHART_DEP_ROOT` should contain your chart
- .helmdocsignore targets `$(BUILDDIR)/helmcharts`, so if you target another path, it won't be taken into account.

**Comment by souleb** [comment]
> once https://github.com/Mellanox/network-operator/pull/1011 is merged, can you followup with a PR to update helm.rst ?

This PR depends on it. Once it will be merged, I'll update `helm.rst` here by running `make branch-helm-docs`

File: Makefile
Comment (adrianchiris) [reply Makefile:22]: can we add support for that ? does it make sense ?

what i was thinking:
1. dev works on some feature in network operator, eventually updates chart and submits PR
2. dev generates updated helm.rst from his local network operator branch and submits PR to network-operator-docs

Comment (souleb) [reply Makefile:22]: the content of the `.helmdocsignore` as to be relative it seems. I wanted to use `realpath` to make it works, but this isn't POSIX and have different behaviour between linux and mac... I don't have a solution so far for this.

Comment (rollandf) [reply Makefile:22]: Maybe have another env PARAM like 'LOCAL_HELM_SOURCE_DIR`, and if it exists, copy content to build/_output/helmcharts/network-operator/charts ?

Comment (souleb) [reply Makefile:22]: let's take some time to think about this one. For now what I think is that a single repo would enable that easily :)

Comment (adrianchiris) [reply Makefile:22]: let me followup with PR on generating from local sources. then each can pick an choose. thx for working on this one @souleb !

**Pull request merged**
